generator/transit_generator.hpp

```diff
@@ -1,9 +1,79 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include "3party/jansson/myjansson.hpp"
+
+#include <cstdint>
 #include <string>
+#include <vector>
 
 namespace routing
 {
 namespace transit
 {
-void BuildTransit(std::string const & mwmPath, std::string const & srtmDir);
+class DeserializerFromJson
+{
+public:
+  DeserializerFromJson(json_struct_t * node) : m_node(node) {}
+
+  void operator()(uint8_t & d, char const * name = nullptr) { GetField(d, name); }
+  void operator()(uint16_t & d, char const * name = nullptr) { GetField(d, name); }
+  void operator()(uint32_t & d, char const * name = nullptr) { GetField(d, name); }
+  void operator()(uint64_t & d, char const * name = nullptr) { GetField(d, name); }
+  void operator()(double & d, char const * name = nullptr) { GetField(d, name); }
+  void operator()(std::string & s, char const * name = nullptr) { GetField(s, name); }
+  void operator()(m2::PointD & p, char const * name = nullptr);
+
+  template <typename R>
+  void operator()(R & r, char const * name = nullptr)
+  {
+    r.Visit(*this);
+  }
+
+  template <typename T>
+  void operator()(std::vector<T> & vs, char const * name = nullptr)
+  {
+    auto * arr = my::GetJSONOptionalField(m_node, name);
+    if (arr == nullptr)
+      return;
+
+    if (!json_is_array(arr))
+      MYTHROW(my::Json::Exception, ("The field", name, "must contain a json array."));
+    size_t const sz = json_array_size(arr);
+    vs.resize(sz);
+    for (size_t i = 0; i < sz; ++i)
+    {
+      DeserializerFromJson arrayItem(json_array_get(arr, i));
+      arrayItem(vs[i]);
+    }
+  }
+
+private:
+  template <typename T>
+  void GetField(T & t, char const * name = nullptr)
+  {
+    if (name == nullptr)
+    {
+      // |name| is not set in case of array items
+      FromJSON(m_node, t);
+    }
+    else
+    {
+      json_struct_t * field = my::GetJSONOptionalField(m_node, name);
+      if (field == nullptr)
```

Comment:
Is it OK that sometimes |t| is not initialized at all, at the end of the method?

Reply:
It's not good but I did it on purpose. It's a case when there's no an optional field at the json. In that case I need to set |t| with a default value. 

Let's consider the class Stop. There's an optional field at it: "TransferId(uint64_t) transfer_id" in it. According to Stop constructor - the default value of |m_transferId| is kTransferIdInvalid (== std::numeric_limits<TransferId>::max()).

At this point I don't know the correct default time for "transfer_id". If set |t| to T() I'll have zero instead of kTransferIdInvalid for "transfer_id".

Reply:
But you call this from several places, and use it not only to parse TransferId. Not good from my point of view.

Reply:
Yes, I call it form several places and use it to parse json to initialize classes like Stop, Header now and will use for classes Shape, Edge and so on in the future. The idea (implemented now) is to set default value for optional json fields at the class which we filled (Stop, Header and so on). 

In example above the default value for Stop::transfer_id is set at class Stop. If it's not changed at the method which called from several places it's left as is.

So if |t| is not initialized in this method the value which set in constractor of class is left.

Probably I did not get the problem you meant. If so, please write in more details.

Reply:
Why don't you return a bool from this method?

Reply:
Yes. I can do that. But how it would help? DeserializerFromJson::GetField() may return bool and DeserializerFromJson::operator() may return bool as well. But I do not see how I can use it then.

Do you think setting default value of optional json fields at corresponding classes is not ok?
Where do you propose to set default value for optional json fields?

Reply:
Default value is fine, but I had the impression that no value was set in some of the code paths.
Though, this fact should be documented anyway.

Reply:
Ok. Now I added a comment at GetField() method:
```
      // No optional field |name| at |m_node|. In that case the default value should be set to |t|.
      // This default value is set at constructor of corresponding class which is filled with
      // |DeserializerFromJson|. And the value (|t|) is not changed at this method.
```
Do you think we should change something else here?